### DISCUSSION TITLE
Lombok 사용 제약사항 도입 제안 (@Data, @Setter, @RequiredArgsConstructor) + @Builder 사용 제안

### DISCUSSION THREAD

**Discussion by Starlight258** [Attica-org/athens-backend#132] [General]
안녕하세요 🙌  팀원 여러분~ 반갑습니다.

Lombok 관련해서 공부하던 중, 몇몇 어노테이션은 사용시 코드 품질을 떨어뜨리고 잠재적으로 버그를 발생시킬 수 있는 위험이 있어서
디스커션을 열게 되었습니다. 
추가적으로 `@RequiredArgsConstructor `사용 관련해서 논의할 부분도 있습니다.

## 1. Lombok 제약사항 도입
@Setter, @AllArgsConstructor, @Data, @Generated 에 관한 내용입니다.
```properties
lombok.Setter.flagUsage = error
lombok.AllArgsConstructor.flagUsage = error
lombok.data.flagUsage = error
lombok.addLombokGeneratedAnnotation = true
```
- `@Setter` 사용 시 오류 발생시키기
  - @Setter 사용시 모든 필드에 대해 변경이 가능합니다. 
  수정되지 않아야하는 값이나 자동으로 넣어지는 값 (ex) `createdAt`)도 수정범위에 들어가므로 좋지 않습니다. 
  필드 수정이 필요할 경우 명시적으로 메서드를 작성하는 것이 코드 의도가 잘 나타날 것 같습니다.

- `@AllArgsConstructor` 사용 시 오류 발생시키기
  - 설정하지 않아야 할 필드(자동 값 생성 필드)까지도 생성자의 값으로 주입받을 위험이 있습니다. 

- `@Data` 사용 시 오류 발생시키기
  - 무분별한 `setter` 남용, `ToString` 양방향 순환 참조, `@EqualsAndHashCode`의 남용 문제가 있습니다.

- Lombok이 생성한 코드에 `@Generated` 어노테이션 추가하기
  - 생성된 코드를 테스트 커버리지에서 제외하여 실제 작성한 코드의 커버리지를 정확히 측정하기 위해서 입니다.

<br/>

## 2. `@RequiredArgsConstructor` 사용에 대한 논의
`@RequiredArgsConstructor`는 final이나 not null 필드에 대해 자동으로 생성자를 만들어 주어 코드 리뷰에서도 제안을 드렸는데요.
관련 글을 읽다보니 해당 어노테이션을 쓰는 것을 지양하는 것이 좋다는 내용을 보게 되었습니다.

동일한 타입의 필드 순서 변경 시 오류가 발생하거나, 의도와 맞지 않게 객체가 생성될 수 있는 위험이 있습니다.
```java
@RequiredArgsConstructor
public class TestUser {
    public String id;
    public String password; // id와 pw 순서가 바뀐다면?
}
```
해당 문제는 생성자에 @Builder를 붙임으로써 필드 순서와 무관하게 안전하게 객체를 생성할 수 있습니다.
```java
public class TestUser {
    public String id;
    public String password;
    private boolean isAdmin;

    @Builder(builderClassName = "UserBuilder", builderMethodName = "normalUserBuilder")
    public TestUser(final String id, final String password) {
        this.id = id;
        this.password = password;
        this.isAdmin = false;
    }

    @Builder(builderClassName = "AdminBuilder", builderMethodName = "adminBuilder")
    private TestUser(final String id, final String password, boolean isAdmin) {
        this.id = id;
        this.password = password;
        this.isAdmin = true;
    }
}
```

```java
// 일반 사용자 생성
TestUser normalUser = TestUser.normalUserBuilder()
                        .id("user123")
                        .password("sadf")
                        .build();

// 관리자 생성
TestUser adminUser = TestUser.adminBuilder()
                        .id("admin")
                        .password("jlkjliw")
                        .build();
```
> `final`을 사용함으로써 값 할당 보장 (값 유효성 검증 코드는 추가적으로 필요)
> 클래스에 `@Builder`는 사용하지 않는 것이 좋다.(`@AllArgsConstructor`와 같은 문제가 발생)
> @Builder에 이름을 붙여 의도와 목적을 표현할 수 있다.

`@RequiredArgsConstructor`를 자주 사용했었는데, 해당 문제를 인지하고 보니 막 사용하기에는 위험이 있네요 😅
위에서 언급드린 Lombok 어노테이션들의 사용을 지양하는 것이 좋을지 여러분의 생각이 궁금합니다!

관련 글은 아래에 첨부해놓았습니다.
감사합니다.

### Reference
[Lombok 을 사용할 때 주의해야 하는 점들 정리](https://jake-seo-dev.tistory.com/70)

@big-cir 
@yybmion

**Comment by big-cir** [comment]
2번에 예제가 @RequiredArgsConstructor
에 해당하는 내용이 맞나요?!
현재는 의존성을 가져오기 위한 용도로 사용하는데, 저희가 사용하는 이유는 예시와는 사용 이유가 다르다고 생각해요.

**Reply by Starlight258** [reply]
@big-cir 님 이야기도 맞습니다.  
주로 의존성 주입 용도로 `@RequiredArgsConstructor`를 사용하는 것이 맞고, 여기서 이야기 드린 부분은 도메인 생성시 생성자를 사용할때 `@RequiredArgsConstructor`를 사용하는것이 좋을지에 대한 내용이에요.
제가 사용 여부를 이야기할 때 도메인 생성 관련 내용인지, 의존성 주입 부분인지 확실히 이야기를 드렸으면 명확했을 것 같네요! 😅

ErrorResponse와 Enum 클래스에서 `@RequiredArgsConstructor`가 자주 사용되고 있어요!

```java
ErrorResponse response = new ErrorResponse(
              ErrorCode.VALIDATION_FAILED.getCode(),
              message
      );
```

**Reply by big-cir** [reply]
아하! 감사합니다!! 제가 너무 확대해석 했나봐요!! 🥲

**Reply by Starlight258** [reply]
앗 아니에요 ㅋㅋㅋㅋ 제가 헷갈리게 써놔서 그랬어요!! 🥹
도메인 생성시에 빌더 패턴을 쓰면 좋을지 계속 이야기해보면 좋을 것 같아요!

사실 빌더 패턴을 사용하면 불안전하게 객체가 생성되어서 사용을 안했었거든요,,
그런데 `Objects.requireNotNull`을 쓰면 필드 필수값이 주입되지 않으면 에러가 발생하게 할 수 있더라구요..!
```java
@Builder
private TestUser(String id, String password) {
    this.id = Objects.requireNonNull(id, "ID must not be null");
    this.password = Objects.requireNonNull(password, "Password must not be null");
}
```
그런데 이게 처음부터 생성자처럼 값 주입이 안되면 컴파일 에러가 나는게 아니라, 생성 후에 런타임 에러로 필수값 주입 여부를 알 수 있어서 그게 좀 걸리긴 합니다,,

생성자가 나을까요 빌더를 사용하는게 나을까요 😂
너무 정답을 찾지 않으려고하는데,, 프로젝트에 적용시에 더 좋은 방법이 무엇일지는 궁금하네요!!

**Reply by big-cir** [reply]
어떤 방법이 좋은지는 정답이 없고 상황마다 다르다고 생각합니다. 어떤 상황에서 적합한지 잘 모르겠을 때에는 아무거나 써보고 좋지 않다고 느껴지면 변경을 해도 되니 어떻게 사용해도 괜찮을거 같아요~!

**Reply by Starlight258** [reply]
ㅋㅋㅋ좋네요 알겠습니다!! 저도 해당 주제에 대해서 좀 더 알아보고 어느정도 기준이 생기면 공유드릴게요!!

**Comment by yybmion** [comment]
`@RequiredAgrsContructor` 어노테이션은 **코드 간결성은 높여 반복적인 코드를 줄이기**에 매우 유용합니다. 
말씀 해주신 것 처럼 
1. 동일한 타입의 **필드 순서 변경**은 생성자에 동일한 타입의 인자가 여러 개 있을 때 문제가 될 수 있다고 봅니다.
2. `@Builder` 패턴, 따라서 위와 같은 상황을 방지하기 위해 빌더 패턴을 사용하면 **필드의 순서와 무관하게 안전하게 객체를 생성**할 수 있습니다. 

2-1. 하지만 이러한 장점도 있지만 @Builder 패턴을 사용하면 **코드의 복잡성이 증가할 수 있어 작은 클래스나 간단한 객체 생성**은 `@RequiredArgsConstructor`를 사용해도 된다고 봅니다.

따라서 상황에 맞게 적절한 어노테이션을 선택하는 것이 좋다고 생각합니다.

**Reply by Starlight258** [reply]
답변 감사합니다 👍
빌더패턴을 사용하면 코드의 복잡성이 증가하신다고 하셨는데, 코드의 양이 많아지는 거 외에 어떤 복잡성이 있는지 설명해주실 수 있나요?
그리고 작은 클래스에서도 동일한 타입의 필드가 존재할 수 있지 않을까요?

맞습니다, 상황에 맞게 적절하게 선택하는 것이 중요합니다,, 기준이 있으면 좋겠네요 🤔

**Reply by yybmion** [reply]
다음 회의 때 이 부분에 대해 토론을 해보는 것이 좋겠네요 👍

**Comment by Starlight258** [comment]
빌더 패턴에서 필수 매개변수는 생성자나 정적 팩토리 메서드를 통해 강제하고, 선택적 매개변수는 원하는대로 설정할수도 있네요!
```java
@Value
@Builder
@Component
public class UserService {

    UserRepository userRepository;  // 필수 매개변수(의존성)

    @Builder.Default
    String defaultUsername = "guest";  // 선택적 매개변수

    @Builder.Default
    int maxUsers = 100;  // 선택적 매개변수

    @Autowired
    public UserService(UserRepository userRepository) {
        this(userRepository, "guest", 100);
    }
}

@Configuration
class AppConfig {

    // 예시 코드
    @Bean
    public UserService userService(UserRepository userRepository) {
        return UserService.builder()
                .userRepository(userRepository)
                .defaultUsername("admin")
                .maxUsers(1000)
                .build();
    }
}
```
